chaincode/src/balances/fetchBalancesWithTokenMetadata.spec.ts

```diff
@@ -0,0 +1,123 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  FetchBalancesDto,
+  FetchBalancesWithTokenMetadataResponse,
+  GalaChainResponse,
+  TokenBalanceWithMetadata,
+  createValidDTO
+} from "@gala-chain/api";
+import { currency, fixture, nft } from "@gala-chain/test";
+import { plainToInstance } from "class-transformer";
+
+import GalaChainTokenContract from "../__test__/GalaChainTokenContract";
+
+it("should Fetch Token Balances with Token Class Metadata", async () => {
+  // Given
+  const currencyClass = currency.tokenClass();
+  const currencyBalance = currency.tokenBalance();
+  const nftClass = nft.tokenClass();
+  const nftBalance = nft.tokenBalance();
+
+  const { ctx, contract } = fixture(GalaChainTokenContract).savedState(
+    currencyClass,
+    nftClass,
+    currencyBalance,
+    nftBalance
+  );
+
+  const dto: FetchBalancesDto = await createValidDTO(FetchBalancesDto, {
+    owner: currency.tokenBalance().owner
+  });
+
+  const curencyBalanceWithMetadata = plainToInstance(TokenBalanceWithMetadata, {
+    balance: currencyBalance,
+    token: currencyClass
+  });
+
+  const nftBalanceWithMetadata = plainToInstance(TokenBalanceWithMetadata, {
+    balance: nftBalance,
+    token: nftClass
+  });
+
+  const expectedResponse = await createValidDTO(FetchBalancesWithTokenMetadataResponse, {
+    results: [curencyBalanceWithMetadata, nftBalanceWithMetadata],
+    nextPageBookmark: ""
+  });
+
+  // When
+  const response = await contract.FetchBalancesWithTokenMetadata(ctx, dto);
+
+  // Then
+  expect(response).toEqual(GalaChainResponse.Success(expectedResponse));
+});
+
+it("should validate the response DTO if the nextPageBookmark is the empty string", async () => {
+  // Given
+  const currencyClass = currency.tokenClass();
+  const nftClass = nft.tokenClass();
+
+  const curencyBalanceWithMetadata = plainToInstance(TokenBalanceWithMetadata, {
+    balance: currency.tokenBalance(),
+    token: currencyClass
+  });
+
+  const nftBalanceWithMetadata = plainToInstance(TokenBalanceWithMetadata, {
+    balance: nft.tokenBalance(),
+    token: nftClass
+  });
+
+  const dto = await createValidDTO(FetchBalancesWithTokenMetadataResponse, {
+    results: [curencyBalanceWithMetadata, nftBalanceWithMetadata],
+    nextPageBookmark: ""
+  });
+
+  // Then
+  const dtoIsValid = await dto
+    .validateOrReject()
+    .then(() => true)
+    .catch((e) => e);
+
+  expect(dtoIsValid).toBe(true);
```

Comment:
optional suggestion: you can also do something like:

```suggestion
  // Then
  await expect(dto.validateOrReject()).resolves.not.toThrow();
```

Reply:
Interesting, I haven't used that syntax before...

I wrote it the original way hoping that if/when the test fails, the jest error message would have a direct expected/received comparison, and a concise error message. Something like, `expected: true \n received: <the caught error message I want to see>`, with the nice green/red contrast colored-text highlighting as seen in a default Linux terminal.

I tested a few scenarios with the `expect().resolves.not.toThrow()` and made some interesting findings. 

Ultimately, I made some changes mostly based on remembering/re-discovering that `createValidDTO` runs the `validateOrReject` function internally. 

In terms of comparing the jest output of different options:

When the error is not caught (all output has uniform default text color):

```
  ● should validate the response DTO if the nextPageBookmark is the empty string

    Property 'nextPageBookmark' of FetchBalancesWithTokenMetadataResponse has failed the following constraints: isNotEmpty

      29 |     const validationErrors = await dto.validate();
      30 |     if (validationErrors.length) {
    > 31 |         throw new DtoValidationFailedError((0, utils_1.getValidationErrorInfo)(validationErrors));
         |               ^
      32 |     }
      33 |     else {
      34 |         return dto;

      at validateDTO (../chain-api/lib/src/types/dtos.js:31:15)
      at async createValidDTO (../chain-api/lib/src/types/dtos.js:52:5)
      at async Object.<anonymous> (src/balances/fetchBalancesWithTokenMetadata.spec.ts:83:10)
```

When the error is caught, and is compared to true (expected line green, received line red, rest default text color):

```
  ● should validate the response DTO if the nextPageBookmark is the empty string

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: [Error: Property 'nextPageBookmark' of FetchBalancesWithTokenMetadataResponse has failed the following constraints: isNotEmpty]

      89 |
      90 |   // Then
    > 91 |   expect(dtoIsValid).toBe(true);
         |                      ^
      92 | });
      93 |
      94 | it("should validate the response DTO if the nextPageBookmark is undefined", async () => {

      at Object.<anonymous> (src/balances/fetchBalancesWithTokenMetadata.spec.ts:91:22)
```

When it fails with the test assertion being that the promise does not throw (I was surprised at how verbose jest made this error output in particular):

```
$ npm run test -- src/balances/fetchBalancesWithTokenMetadata.spec.ts 

> @gala-chain/chaincode@1.2.4 test
> jest src/balances/fetchBalancesWithTokenMetadata.spec.ts

<path>/node_modules/expect/build/index.js:174
    const innerErr = new JestAssertionError();
                     ^
JestAssertionError: expect(received).resolves.not.toThrow()

Error name:    "Error"
Error message: "Property 'nextPageBookmark' of FetchBalancesWithTokenMetadataResponse has failed the following constraints: isNotEmpty"

      29 |     const validationErrors = await dto.validate();
      30 |     if (validationErrors.length) {
    > 31 |         throw new DtoValidationFailedError((0, utils_1.getValidationErrorInfo)(validationErrors));
         |               ^
      32 |     }
      33 |     else {
      34 |         return dto;

      at validateDTO (../chain-api/lib/src/types/dtos.js:31:15)
      at async createValidDTO (../chain-api/lib/src/types/dtos.js:52:5)
    at Object.toThrow (./sdk/node_modules/expect/build/index.js:174:22)
    at Object.<anonymous> (./sdk/chaincode/src/balances/fetchBalancesWithTokenMetadata.spec.ts:89:28)
    at Promise.then.completed (./sdk/node_modules/jest-circus/build/utils.js:298:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (./sdk/node_modules/jest-circus/build/utils.js:231:10)
    at _callCircusTest (./sdk/node_modules/jest-circus/build/run.js:316:40)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _runTest (./sdk/node_modules/jest-circus/build/run.js:252:3)
    at async _runTestsForDescribeBlock (./sdk/node_modules/jest-circus/build/run.js:126:9)
    at async run (./sdk/node_modules/jest-circus/build/run.js:71:3)
    at async runAndTransformResultsToJestFormat (./sdk/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapterInit.js:122:21)
    at async jestAdapter (./sdk/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:79:19)
    at async runTestInternal (./sdk/node_modules/jest-runner/build/runTest.js:367:16)
    at async runTest (./sdk/node_modules/jest-runner/build/runTest.js:444:34) {
  matcherResult: {
    message: '\x1B[2mexpect(\x1B[22m\x1B[31mreceived\x1B[39m\x1B[2m).\x1B[22mresolves\x1B[2m.\x1B[22mnot\x1B[2m.\x1B[22mtoThrow\x1B[2m()\x1B[22m\n' +
      '\n' +
      'Error name:    \x1B[31m"Error"\x1B[39m\n' +
      `Error message: \x1B[31m"Property 'nextPageBookmark' of FetchBalancesWithTokenMetadataResponse has failed the following constraints: isNotEmpty"\x1B[39m\n` +
      '\n' +
      '    \x1B[0m \x1B[90m 29 |\x1B[39m     \x1B[36mconst\x1B[39m validationErrors \x1B[33m=\x1B[39m \x1B[36mawait\x1B[39m dto\x1B[33m.\x1B[39mvalidate()\x1B[33m;\x1B[39m\n' +
      '     \x1B[90m 30 |\x1B[39m     \x1B[36mif\x1B[39m (validationErrors\x1B[33m.\x1B[39mlength) {\n' +
      '    \x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 31 |\x1B[39m         \x1B[36mthrow\x1B[39m \x1B[36mnew\x1B[39m \x1B[33mDtoValidationFailedError\x1B[39m((\x1B[35m0\x1B[39m\x1B[33m,\x1B[39m utils_1\x1B[33m.\x1B[39mgetValidationErrorInfo)(validationErrors))\x1B[33m;\x1B[39m\n' +
      '     \x1B[90m    |\x1B[39m               \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\n' +
      '     \x1B[90m 32 |\x1B[39m     }\n' +
      '     \x1B[90m 33 |\x1B[39m     \x1B[36melse\x1B[39m {\n' +
      '     \x1B[90m 34 |\x1B[39m         \x1B[36mreturn\x1B[39m dto\x1B[33m;\x1B[39m\x1B[0m\n' +
      '\n' +
      '      \x1B[2mat validateDTO (\x1B[22m../chain-api/lib/src/types/dtos.js\x1B[2m:31:15)\x1B[22m\n' +
      '      \x1B[2mat async createValidDTO (\x1B[22m../chain-api/lib/src/types/dtos.js\x1B[2m:52:5)\x1B[22m',
    pass: true
  }
}
```

The above pastes into the Github comment don't capture the syntax highlighting - when comparing expected/received, the one line I most want to read gets the red error highlighting. 

I decided to go with the simplest error output and clearest highlighting. I pushed a small commit that shortens the tests by just using the `createValidDTO` function instead of redundantly calling `verifyOrReject()`. 

Jest is complicated, it is always useful to learn more about its handling of different scenarios.